### PR TITLE
Fix: Collateral test failures in test_versioning (#2611)

### PR DESCRIPTION
**Pull request by amarjandu**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [ ] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Purchased BQ slot committment in `dev`                <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Purchased BQ slot committment in `prod`               <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
Commit and PR titles are non-sensical.

File: deployments/dev/environment.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def env() -> Mapping[str, Optional[str]]:
             'hca:it2ebi:repository/tdr:metadata/hca'
         ]),
 
-        'AZUL_TDR_SOURCES': ','.join([
+        'AZUL_TDR_SOURCE': ','.join([
```
Comment (noah-aviel-dove) [inline deployments/dev/environment.py:43]: Was this change intentional? It seems unrelated and breaks the build

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]
Build is red.

File: deployments/dev/environment.py (hunk 1 above)
Comment (amarjandu) [reply deployments/dev/environment.py:43]: Yeah its in a drop commit to show how/why things are failing, this failure shows that the failure is now contained to a single test case, and does not spread to other tests.

**Comment by amarjandu** [comment]
> Build is red.

Yeah for the drop commit it is red.

File: test/service/test_portal_service.py [hunk 2]
```diff
@@ -107,15 +107,14 @@ def tearDown(self):
...
+        # must be deleted.
         self.s3_client.put_bucket_versioning(Bucket=self.portal_service.bucket,
                                              VersioningConfiguration={
                                                  'Status': 'Disabled',
                                                  'MFADelete': 'Disabled'
                                              })
-        self.s3_client.delete_object(Bucket=self.portal_service.bucket,
-                                     Key=self.portal_service.object_key)
+        bucket = aws.resource('s3').Bucket(self.portal_service.bucket)
+        bucket.objects.all().delete()
```
Comment (noah-aviel-dove) [inline test/service/test_portal_service.py:117]: If the intent here is to avoid raising an exception if the target object does not exist, surely a simple try/catch would be simpler, clearer, and not run the risk of causing unintended deletions? Otherwise I don't understand the reason for this change.

File: deployments/dev/environment.py (hunk 1 above)
Comment (noah-aviel-dove) [reply deployments/dev/environment.py:43]: I'm still not clear on what this commit is supposed to demonstrate. You changed the name of an environment variable and [observed](https://travis-ci.com/github/DataBiosphere/azul/jobs/482481471) the obvious consequence, a `KeyError` when the configuration tries to read that variable. I don't see any other errors. How then, can this failure be considered "collateral"?

If you're just trying to show that the environment mis-configuration _doesn't_ cascade into other test failures, then the commit title is wrong.

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]

File: deployments/dev/environment.py (hunk 1 above)
Comment (hannes-ucsc) [reply deployments/dev/environment.py:43]: @amarjandu, this is a very roundabout way of causing the tests to fail. I can understand Noah's confusion.

Why don't you just raise an explicit exception in the test or tests that you want to show cause collateral failures of other tests?

**Comment by amarjandu** [comment]
@hannes-ucsc I added the drop commits that you wanted to see per our discussion yesterday after standup: 

1. Assert False in a test to see a single failure.
2. Assert False in the cleanup to see a collateral failure within another test (and class in our case)

I introduced a fix for this issue that resets the moto backend when the tests are setup, which will pass if the `drop!` commits are removed. If I had introduced the fix commit before the `drop!` commits, the collateral failure would not be visible, it would just show that each test fails due to an assert false.

I think it would be strange to request a review from @noah-aviel-dove without a passing build on the commit that address the issue. Would it be OK to remove the drop commits for the reproduction of the collateral failure and proceed with the review?

**Comment by hannes-ucsc** [comment]
With the more immediate and explicit technique for failing test and teardown, it's much more clear what your trying to demonstrate. If the peer reviewer doesn't understand that, they shouldn't be reviewing this PR. We will remove the drop commits as part of my review.

**Comment by hannes-ucsc** [comment]
@amarjandu, one thing you could add is yet another drop! commit that reverts the previous drop! commits. That commit would be expected to have green status.

File: test/azul_test_case.py [hunk 3]
```diff
@@ -344,3 +347,18 @@ def setUp(self):
...
+    """
+
+    def setUp(self) -> None:
+        super().setUp()
+        # Moto backends are reset to ensure no residual resources are left over if
+        # a test fails to cleanup after itself.
+        moto_api_backend.reset()
+
+    def tearDown(self) -> None:
+        super().tearDown()
```
Comment (noah-aviel-dove) [inline test/azul_test_case.py:364]: This seems unnecessary

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]
You reference the issue in two different commit titles. I don't believe this is correct. I would remove it from "Add super setup/teardown to VersionTableTestCase" since this commit doesn't directly address the underlying issue.

Reproduction and resolution of collaterals looks good.

File: test/azul_test_case.py (hunk 3 above)
Comment (amarjandu) [reply test/azul_test_case.py]: > You reference the issue in two different commit titles. I don't believe this is correct. I would remove it from "Add super setup/teardown to VersionTableTestCase" since this commit doesn't directly address the underlying issue.

Thanks for catching that, I had first added the moto backend resets as a part of the teardown for the AzulUnitTestCase, so it was necessary for the fix. I've updated the commit titles.

**Comment by amarjandu** [comment]
I'll get all the commits before requesting a primary review.

**Review by noah-aviel-dove — APPROVED** [review]
Looks good to me!

File: test/azul_test_case.py [hunk 4]
```diff
@@ -349,3 +352,15 @@ def setUp(self):
                 with self.assertRaises(KeyboardInterrupt):
                     TC(events).run()
                 self.assertEqual(events, expected_events)
+
+
+class AzulMotoTestCase(TestCase):
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:357]: How do we ensure that someone doesn't introduce Moto mocks to their test but forgets to mix this in?

File: test/azul_test_case.py [hunk 5]
```diff
@@ -349,3 +352,15 @@ def setUp(self):
...
+    """
+    A mixin for tests that utilize the moto module
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:359]:
```suggestion
    A mixin for tests that utilize the Moto library.
```

File: test/azul_test_case.py [hunk 6]
```diff
@@ -349,3 +352,15 @@ def setUp(self):
...
+    """
+
+    def setUp(self) -> None:
+        super().setUp()
+        # Moto backends are reset to ensure no residual resources are left over if
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:364]:
```suggestion
        # Moto backends are reset to ensure no resources are left over if
```

Residual means left-over.

File: test/azul_test_case.py [hunk 7]
```diff
@@ -349,3 +352,15 @@ def setUp(self):
...
+        # a test fails to cleanup after itself.
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:365]:
```suggestion
        # a test fails to clean up after itself.
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
No bump.

File: test/version_table_test_case.py [hunk 8]
```diff
@@ -24,6 +24,7 @@ class VersionTableTestCase(AzulUnitTestCase):
     _aws_test_region = 'ap-south-1'
 
     def setUp(self):
+        super().setUp()
```
Comment (hannes-ucsc) [inline test/version_table_test_case.py:27]: oops

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
For demo, replicate the experiment with the drop commits.

**Comment by jessebrennan** [comment]
Integration test in sandbox failed. I'm trying again https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/10794

**Comment by amarjandu** [comment]
@jessebrennan very strange, i just completed the IT on my personal deployment without an issue.

**Comment by amarjandu** [comment]
From the logs it looks like the issue was with the the IAM policy that the lambda was using, might have been a deployment problem.

File: test/azul_test_case.py [hunk 9]
```diff
@@ -149,6 +149,15 @@ def tearDownClass(cls) -> None:
         cls._restore_catalogs()
         super().tearDownClass()
 
+    def setUp(self) -> None:
```
Comment (amarjandu) [inline test/azul_test_case.py:152]: add comment 
```
FIXME: remove local import and update moto 
             #1718 
```

Comment (amarjandu) [reply test/azul_test_case.py:152]: full link plz

Comment (amarjandu) [reply test/azul_test_case.py:152]: Commit Title: 
```
Work around spulec/moto#2058 (#1718)
```

**Review by hannes-ucsc — APPROVED** [review]
